sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -134,49 +155,81 @@ case class CreateMap(children: Seq[Expression]) extends Expression {
   }
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    val arrayClass = classOf[GenericArrayData].getName
     val mapClass = classOf[ArrayBasedMapData].getName
     val keyArray = ctx.freshName("keyArray")
     val valueArray = ctx.freshName("valueArray")
-    ctx.addMutableState("Object[]", keyArray, s"this.$keyArray = null;")
-    ctx.addMutableState("Object[]", valueArray, s"this.$valueArray = null;")
 
-    val keyData = s"new $arrayClass($keyArray)"
-    val valueData = s"new $arrayClass($valueArray)"
+    val MapType(keyDt, valueDt, _) = dataType
+    val evalKeys = keys.map(e => e.genCode(ctx))
+    val isPrimitiveArrayKey = ctx.isPrimitiveType(keyDt) && evalKeys.forall(_.isNull == "false")
+    val evalValues = values.map(e => e.genCode(ctx))
+    val isPrimitiveArrayValue =
+      ctx.isPrimitiveType(keyDt) && evalValues.forall(_.isNull == "false")
+
+    val (keyData, keyArrayAllocate, keyArrayNullify) = if (!isPrimitiveArrayKey) {
```

Comment:
These two blocks to generate codes for key and value look very similar.

Can we abstract a private method to do this and use parameters to decide generating key or value array codes?